category_theory/opposites.lean

```diff
@@ -21,34 +21,77 @@ instance opposite : category.{u₁ v₁} (Cᵒᵖ) :=
   comp := λ _ _ _ f g, g ≫ f,
   id   := λ X, 𝟙 X }
 
+def op_op : (Cᵒᵖ)ᵒᵖ ⥤ C :=
+{ obj := λ X, X,
+  map := λ X Y f, f }
+
+-- TODO this is an equivalence
+
 namespace functor
 
 section
 variables {D : Type u₂} [𝒟 : category.{u₂ v₂} D]
 include 𝒟
 
-protected definition op (F : C ⥤ D) : (Cᵒᵖ) ⥤ (Dᵒᵖ) :=
-{ obj       := λ X, F.obj X,
-  map       := λ X Y f, F.map f,
-  map_id'   := begin /- `obviously'` says: -/ intros, erw [map_id], refl, end,
-  map_comp' := begin /- `obviously'` says: -/ intros, erw [map_comp], refl end }
+variables (C D)
+
+definition op_hom : (C ⥤ D)ᵒᵖ ⥤ (Cᵒᵖ ⥤ Dᵒᵖ) :=
+{ obj := λ F : C ⥤ D,
+  { obj       := λ X, F.obj X,
+    map       := λ X Y f, F.map f,
+    map_id'   := begin /- `obviously'` says: -/ intros, erw [map_id], refl, end,
+    map_comp' := begin /- `obviously'` says: -/ intros, erw [map_comp], refl end },
+  map := λ F G α,
+  { app := λ X, α.app X,
+    naturality' := λ X Y f, eq.symm (α.naturality f) } }
+
+@[simp] lemma op_hom.map_app {F G : (C ⥤ D)ᵒᵖ} (α : F ⟶ G) (X : C) :
+  ((op_hom C D).map α).app X = α.app X := rfl
+
+definition op_inv : (Cᵒᵖ ⥤ Dᵒᵖ) ⥤ (C ⥤ D)ᵒᵖ :=
+{ obj := λ F : Cᵒᵖ ⥤ Dᵒᵖ,
+  { obj       := λ X, F.obj X,
+    map       := λ X Y f, F.map f,
+    map_id'   := F.map_id,
+    map_comp' := by intros; apply F.map_comp },
+  map := λ F G α,
+  { app := λ X : C, α.app X,
+    naturality' := λ X Y f, eq.symm (α.naturality f) } }
+
+@[simp] lemma op_inv.obj_obj (F : Cᵒᵖ ⥤ Dᵒᵖ) (X : C) : ((op_inv C D).obj F).obj X = F.obj X := rfl
+@[simp] lemma op_inv.obj_map (F : Cᵒᵖ ⥤ Dᵒᵖ) {X Y : C} (f : X ⟶ Y) :
+  ((op_inv C D).obj F).map f = F.map f := rfl
+@[simp] lemma op_inv.map_app {F G : Cᵒᵖ ⥤ Dᵒᵖ} (α : F ⟶ G) (X : C) :
+  ((op_inv C D).map α).app X = α.app X := rfl
+
+-- TODO show these form an equivalence
+
+variables {C D}
+
+protected definition op (F : C ⥤ D) : Cᵒᵖ ⥤ Dᵒᵖ := (@op_hom C _ D _).obj (F : (C ⥤ D)ᵒᵖ)
```

Comment:
I think when you are going to define both the higher order functor and the object part, you should do it in the opposite order, i.e. copy the definition of `op_hom.obj` in here and use this definition in the op field of `op_hom`. That way there is less unfolding and `op_hom` unfolds to `op` like you would expect.

Reply:
Thanks, I've fixed this.